File: [P3] estrutura de dados/UNIDADE 2/TABELA HASH/hash.py
class HashTable:
    
    def __init__(self, n):
        self.n = n
        self.table = [[] for _ in range(10)]

    def hash1(self, c):
        return c % 10

    def hash2(self, c):
        return c % (self.n // 10)

    def insert(self, c):
        index = self.hash1(c)
        self.table[index].append(c)

    def remove(self, c):
        index = self.hash1(c)
        self.table[index].remove(c)

    def search(self, c):
        index = self.hash1(c)
        for i in self.table[index]:
            if i == c:
                return True
        return False

ht = HashTable(100)

# Inserindo elementos
ht.insert(10)
ht.insert(20)
ht.insert(15)

# Procurando elementos
print("Procurando por 10:", ht.search(10))
print("Procurando por 20:", ht.search(20))
print("Procurando por 15:", ht.search(15))

# Removendo elementos
ht.remove(20)

# Procurando elementos após a remoção
print("Procurando por 20:", ht.search(20))

File: [P3] estrutura de dados/UNIDADE 2/TABELA HASH/hash.cpp
#include <iostream>
#include <vector>
#include <list>

using namespace std;

const int n = 101; // Tamanho da tabela principal

// Função de hash para o primeiro nível
int FirstLevelHash(int key) {
    return key % 10;
}

// Função de hash para o segundo nível
int SecondLevelHash(int key) {
    return key % (n / 10);
}

class HashTable {
private:
    vector<list<int>> firstLevelTable[10];
    vector<list<int>> secondLevelTable[n / 10];

public:
    // Insere um objeto com uma chave na tabela hash
    void Insert(int key) {
        int firstIndex = FirstLevelHash(key);
        int secondIndex = SecondLevelHash(key);

        firstLevelTable[firstIndex][secondIndex].push_back(key);
    }

    // Procura um objeto na tabela hash
    bool Search(int key) {
        int firstIndex = FirstLevelHash(key);
        int secondIndex = SecondLevelHash(key);

        for (int value : firstLevelTable[firstIndex][secondIndex]) {
            if (value == key) {
                return true;
            }
        }
        return false;
    }
};

int main() {
    HashTable hashTable;

    // Inserindo valores de exemplo
    hashTable.Insert(10);
    hashTable.Insert(25);
    hashTable.Insert(35);

    // Procurando valores de exemplo
    cout << "Procurando 10: " << (hashTable.Search(10) ? "Encontrado" : "Não encontrado") << endl;
    cout << "Procurando 25: " << (hashTable.Search(25) ? "Encontrado" : "Não encontrado") << endl;
    cout << "Procurando 42: " << (hashTable.Search(42) ? "Encontrado" : "Não encontrado") << endl;

    return 0;
}

